internal/nmc/module_nmc_controller.go

```diff
@@ -0,0 +1,142 @@
+package nmc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/api"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//+kubebuilder:rbac:groups="core",resources=nodes,verbs=get;watch;patch
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs,verbs=get;list;watch;update;patch;create
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs/status,verbs=get;update;patch
+
+type ModuleNMCHandler interface {
+	handleModuleNMC(ctx context.Context, mld *api.ModuleLoaderData, nodes []v1.Node) error
+}
+
+type moduleNMCHandler struct {
+	reconHelper moduleNMCHandlerHelperAPI
+}
+
+func NewModuleNMCHandler(client client.Client,
+	nmcHelper Helper) ModuleNMCHandler {
+	reconHelper := newModuleNMCHandlerHelper(client, nmcHelper)
+	return &moduleNMCHandler{
+		reconHelper: reconHelper,
+	}
+}
+
+// func (mnr *ModuleNMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (mnh *moduleNMCHandler) handleModuleNMC(ctx context.Context, mld *api.ModuleLoaderData, nodes []v1.Node) error {
+	var sumErr error
+	for _, node := range nodes {
+		nodeKernelVersion := strings.TrimSuffix(node.Status.NodeInfo.KernelVersion, "+")
+
+		shouldBeOnNode, err := mnh.reconHelper.shouldModuleRunOnNode(node, mld, nodeKernelVersion)
+		if err != nil {
+			sumErr = errors.Join(sumErr, err)
+			continue
+		}
+		if shouldBeOnNode {
+			err = mnh.reconHelper.enableModuleOnNode(ctx, mld, node.Name, nodeKernelVersion)
+		} else {
+			err = mnh.reconHelper.disableModuleOnNode(ctx, mld.Namespace, mld.Name, node.Name)
+		}
+
+		sumErr = errors.Join(sumErr, err)
```

Comment:
This is not the recommended pattern: https://go-review.googlesource.com/c/go/+/495076
Instead, let us append to a `[]error` and use `errors.Join(errs...)` at the end of the loop.

Reply:
fixed